### DISCUSSION TITLE
Add user Phone Number field and Phone Number Locale, with validation and verification capabilities.

### DISCUSSION THREAD

**Discussion by Yejneshwar** [keycloak/keycloak#15012] [Ideas]
I have been working on this feature on my fork.

# Proposal

libphonenumber can be used for validating phone numbers, and SMS providers can be added for verification of said phone numbers.

SMS can be used just to verify phone numbers.

Although not recommended, keycloak can also be configured to use SMS for 2FA (maybe not implemented at all?).  

# Reason

Some payment providers (e.g. juspay) and other service providers, need a valid phone number to create a customer object on said service. Keycloak could provide built-in functionality to store, update, validate & verify user phone numbers.

# UI

![Capture10](https://user-images.githubusercontent.com/81953958/196663658-7a47bec7-51f1-4328-bc01-e29c2a3bccca.PNG)

**Comment by pengfan8080** [comment]
Can you provide detailed configuration screenshots, or a small demo for mobile number verification

**Reply by Yejneshwar** [reply]
In the screenshot above verification had been disabled for testing purposes.
Will update soon

**Comment by d3bt3ch** [comment]
@Yejneshwar Can you write a custom `authenticator` and share with the community please. You can use [authentication service provider interface](https://www.keycloak.org/docs/latest/server_development/#_auth_spi)

**Reply by Yejneshwar** [reply]
Almost all the work is done, just need to work on setting up a GUI to configure endpoints for SMS API's (Phone number verification). probs the AWS pinpoint service?

**Reply by d3bt3ch** [reply]
Great. Can you please share the existing progress.

**Reply by Yejneshwar** [reply]
Take a look at [Yej/keycloak](https://github.com/Yejneshwar/keycloak/tree/error-branch) should build as usual.

build and Install using `mvn clean install` [Yej/Keycloak-ui](https://github.com/Yejneshwar/keycloak-ui) first.

To verify phone numbers turn on `Verify Phone Numer` in Realm settings.

Use the `UPDATE_PHONE_NUMBER` required action.

**Comment by ryan-morris** [comment]
@Yejneshwar is this something you are still working on? Was your intention to have it all added as an extension or to PR and bring into the main project?

**Comment by Yejneshwar** [comment]
The aim is to bring it to the main project. I'm stuck on how to implement the various SMS providers for phone number verification. Maybe just the SMS providers are extensions? or just cover the major providers? and any additional providers are extensions? any ideas are welcome.
